clusterversion: benchmark and rm Unmarshal allocs

pav-kv opened the pull request:
Currently, each `IsActive` does a memory allocation:

```
==================== Test output for //pkg/clusterversion:clusterversion_test:
goos: darwin
goarch: arm64
BenchmarkClusterVersionSettingIsActive
BenchmarkClusterVersionSettingIsActive-10       28778041                42.03 ns/op           16 B/op          1 allocs/op
PASS
```

Since the cluster version check is in many hot paths, we should eliminate this allocation.

After:

```
==================== Test output for //pkg/clusterversion:clusterversion_test:
goos: darwin
goarch: arm64
BenchmarkClusterVersionSettingIsActive
BenchmarkClusterVersionSettingIsActive-10       45417914                26.43 ns/op            0 B/op          0 allocs/op
PASS
```

Touches #111561
Epic: none
Release note: none

pav-kv committed:
clusterversion: benchmark IsActive version check

Currently, each IsActive does a memory allocation:

```
==================== Test output for //pkg/clusterversion:clusterversion_test:
goos: darwin
goarch: arm64
BenchmarkClusterVersionSettingIsActive
BenchmarkClusterVersionSettingIsActive-10       28778041                42.03 ns/op           16 B/op          1 allocs/op
PASS
```

Since the cluster version check is in many hot paths, we should
eliminate this allocation.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/113043)
<!-- Reviewable:end -->

pavelkalinnikov commented:
After some experiments, it seems like the allocation is hidden inside `protoutil.Unmarshal`.

pav-kv committed:
clusterversion: rm alloc in activeVersionOrEmpty

The allocation no longer presents itself in the benchmark.

```
==================== Test output for //pkg/clusterversion:clusterversion_test:
goos: darwin
goarch: arm64
BenchmarkClusterVersionSettingIsActive
BenchmarkClusterVersionSettingIsActive-10       45417914                26.43 ns/op            0 B/op          0 allocs/op
PASS
```

Epic: none
Release note: none

pavelkalinnikov commented:
There are more `Unmarshal` calls in this package which probably could be optimized too.

pav-kv committed:
clusterversion: rm allocs in ClusterVersion.Unmarshal

Epic: none
Release note: none

erikgrinaker commented:
> After some experiments, it seems like the allocation is hidden inside `protoutil.Unmarshal`.

Yeah, gogoproto/Protobuf is very inefficient in terms of allocations. We've considered using some other encoding, e.g. Tobi was playing around with FlatBuffers last year, but it hasn't been prioritized. In the short term, it doesn't seem viable to replace it.

erikgrinaker commented:
I think there was a reason why we use `protoutil.Unmarshal()` over `Message.Unmarshal()`, I think there's even a linter for it. I'll have to refresh my memory on this stuff.

pavelkalinnikov commented:
> I think there was a reason why we use `protoutil.Unmarshal()` over `Message.Unmarshal()`, I think there's even a linter for it. I'll have to refresh my memory on this stuff.

Interesting.

When looking at code, I concluded that `protoutil.Unmarshal` pretty much calls `Reset` and `Unmarshal` on the message. We don't need the `Reset` in all these cases because the message is created fresh, so we can just call `Unmarshal` on the concrete message.

pavelkalinnikov commented:
We do have this lint indeed.

erikgrinaker commented:
That's what I thought:

```
    lint_test.go:1295:
        clusterversion/setting.go:140:  if err := curVer.Unmarshal(encoded.([]byte)); err != nil { <- forbidden; use 'protoutil.Unmarshal' instead
    lint_test.go:1295:
        clusterversion/setting.go:157:  if err := clusterVersion.Unmarshal(val); err != nil { <- forbidden; use 'protoutil.Unmarshal' instead
    lint_test.go:1295:
        clusterversion/setting.go:168:  if err := newCV.Unmarshal(newRawProto); err != nil { <- forbidden; use 'protoutil.Unmarshal' instead
    lint_test.go:1295:
        clusterversion/setting.go:177:  if err := oldCV.Unmarshal(curRawProto); err != nil { <- forbidden; use 'protoutil.Unmarshal' instead
    lint_test.go:1295:
        clusterversion/setting.go:213:  if err := ver.Unmarshal(rawProto); err != nil { <- forbidden; use 'protoutil.Unmarshal' instead
```

It's probably okay anyway, but I'll have a closer look at this when I have time.

pavelkalinnikov commented:
The linter was introduced here: #18455. I think it's overly restrictive. It costs unnecessary allocations.

I tried generics:

```golang
func unmarshal[Message protoutil.Message](m Message, b []byte) error {
	m.Reset()
	return m.Unmarshal(b)
}

...

if err := unmarshal[*ClusterVersion](&curVer, encoded.([]byte)); err != nil {
	log.Fatalf(ctx, "%v", err)
}
```
but it doesn't remove the allocation. I don't think the calls are inlined in this case.

So the only thing we can do is either lift or workaround this restriction. Maybe we can silence the linter knowing that this case is safe.

erikgrinaker commented on pkg/clusterversion/setting.go:
```diff
@@ -165,7 +165,7 @@ func (cv *clusterVersionSetting) ValidateVersionUpgrade(
 	_ context.Context, sv *settings.Values, curRawProto, newRawProto []byte,
 ) error {
 	var newCV ClusterVersion
-	if err := protoutil.Unmarshal(newRawProto, &newCV); err != nil {
```
We don't strictly need to change all of these, they're not hot paths. I think `activeVersionOrEmpty`, and possibly `Decode`, are the ones we care about.

dt replied:
+1

Honestly we could also just move clusterversion.proto into another package that pkg/settings is allowed to depend on and then store a *ClusterVersion in the atomic instead of a []byte so there'd be no unmarshal -- allocating or otherwise -- on the hot path. But that's a bigger change, so if you're eyeing a back port to resolve an observed customer issue, this more targeted removal of just the alloc, and just on the hot path, probably makes sense for now.

pavelkalinnikov replied:
Leaving as is so that this file consistently uses one way. LMK if feel strongly.
I will file an issue to get rid of protos in this package.

erikgrinaker reviewed (approved):
>So the only thing we can do is either lift or workaround this restriction. Maybe we can silence the linter knowing that this case is safe.

You can add an exception for `clusterversion/setting.go` here, and add a comment at the call sites for why we don't use `protoutil.Unmarshal()`:

https://github.com/cockroachdb/cockroach/blob/fb37a8215c68482aa0e2934d195d0d5c66f62fbd/pkg/testutils/lint/lint_test.go#L1274-L1281

As @dt pointed out, it seems unfortunate that we're even unmarshaling here, since `isActive()` is commonly used in hot paths. Let's write up an issue for this.

erikgrinaker commented:
Btw, I think we could add a generic `protoutil.UnmarshalNew()` which always creates a new message, guaranteeing that it's safe to omit `Reset()`. Something like:

```go
func UnmarshalNew[M Message](data []byte) (M, error) {
  var m M
  return m.Unmarshal(data)
}
```

pavelkalinnikov commented:
> Btw, I think we could add a generic `protoutil.UnmarshalNew()` which always creates a new message, guaranteeing that it's safe to omit `Reset()`. Something like:
> 
> ```go
> func UnmarshalNew[M Message](data []byte) (M, error) {
>   var m M
>   return m.Unmarshal(data)
> }
> ```

Tried that:

```golang
func UnmarshalNew[M Message](data []byte) (M, error) {
	var m M
	err := m.Unmarshal(data)
	return m, err
}

// and then
curVer, err := protoutil.UnmarshalNew[ClusterVersion](encoded.([]byte))
if err != nil {
	log.Fatalf(ctx, "%v", err)
}
```

This does not build:

```
pkg/clusterversion/setting.go:139:40: ClusterVersion does not satisfy protoutil.Message (method MarshalTo has pointer receiver)
```

Setting the `M` type to pointer doesn't help either: `var m M` becomes a nil, and then there is a nil deref panic.

erikgrinaker commented:
> Setting the `M` type to pointer doesn't help either: `var m M` becomes a nil, and then there is a nil deref panic.

Can do `m := new(M)` or something, perhaps.

Either way, as you pointed out on Slack, this may not help with performance anyway due to how Go monomorphizes generic functions.

pav-kv committed:
lint: allow clusterversion to use proto Unmarshal

The mandated protoutil.Unmarshal caused unnecessary allocations in
clusterversion/setting.go, is used extensively on critical paths. These
allocations were visible on benchmarks (e.g. 1% of all allocations in
one KV benchmark).

Epic: none
Release note: none

pavelkalinnikov commented:
bors retry
